src/ucs/vfs/sock/vfs_sock.c

```diff
@@ -5,7 +5,163 @@
+#include <pwd.h>
+
+
+typedef struct {
+    uint8_t action;
+} UCS_S_PACKED ucs_vfs_msg_t;
+
+
+int ucs_vfs_sock_get_address(struct sockaddr_un *un_addr)
+{
+    struct passwd *pw;
+    int ret;
+
+    pw = getpwuid(geteuid());
+    if (pw == NULL) {
+        return -errno;
+    }
+
+    memset(un_addr, 0, sizeof(*un_addr));
+    un_addr->sun_family = AF_UNIX;
+    snprintf(un_addr->sun_path, sizeof(un_addr->sun_path) - 1,
+             "/tmp/ucx-vfs-%s.sock", pw->pw_name);
+    return 0;
+}
+
+int ucs_vfs_sock_setopt_passcred(int sockfd)
+{
+    int optval, ret;
+
+    optval = 1;
+    ret = setsockopt(sockfd, SOL_SOCKET, SO_PASSCRED, &optval, sizeof(optval));
+    if (ret < 0) {
+        return -errno;
+    }
+
+    return 0;
+}
+
+static int ucs_vfs_sock_retval(ssize_t ret, size_t expected)
+{
+    if (ret == expected) {
+        return 0;
+    } else if (ret < 0) {
+        return -errno;
+    } else {
+        return -EIO;
+    }
+}
+
+int ucs_vfs_sock_send(int sockfd, const ucs_vfs_sock_message_t *vfs_msg)
+{
+    char cbuf[CMSG_SPACE(sizeof(*vfs_msg))] UCS_V_ALIGNED(sizeof(size_t));
+    struct cmsghdr *cmsgp;
+    struct msghdr msgh;
+    struct ucred cred;
+    ucs_vfs_msg_t msg;
+    struct iovec iov;
+    ssize_t nsent;
+
+    memset(cbuf, 0, sizeof(cbuf));
+    memset(&msgh, 0, sizeof(msgh));
+    msg.action      = vfs_msg->action;
+    iov.iov_base    = &msg;
+    iov.iov_len     = sizeof(msg);
+    msgh.msg_iov    = &iov;
+    msgh.msg_iovlen = 1;
+
+    if (vfs_msg->action == UCS_VFS_SOCK_ACTION_MOUNT_REPLY) {
+        /* send file descriptor */
+        msgh.msg_control    = cbuf;
+        msgh.msg_controllen = sizeof(cbuf);
+        cmsgp               = CMSG_FIRSTHDR(&msgh);
+        cmsgp->cmsg_level   = SOL_SOCKET;
+        cmsgp->cmsg_len     = CMSG_LEN(sizeof(vfs_msg->fd));
+        cmsgp->cmsg_type    = SCM_RIGHTS;
+        memcpy(CMSG_DATA(cmsgp), &vfs_msg->fd, sizeof(vfs_msg->fd));
+    }
+
+    do {
+        nsent = sendmsg(sockfd, &msgh, 0);
+    } while ((nsent < 0) & (errno == EINTR));
+    return ucs_vfs_sock_retval(nsent, iov.iov_len);
+}
+
+int ucs_vfs_sock_recv(int sockfd, ucs_vfs_sock_message_t *vfs_msg)
+{
+    char cbuf[CMSG_SPACE(sizeof(*vfs_msg))] UCS_V_ALIGNED(sizeof(size_t));
+    struct cmsghdr *cmsgp;
+    struct msghdr msgh;
+    struct ucred cred;
+    ucs_vfs_msg_t msg;
+    struct iovec iov;
+    ssize_t nrecvd;
+
+    /* initialize to invalid values */
+    vfs_msg->action = UCS_VFS_SOCK_ACTION_INVALID;
+    vfs_msg->fd     = -1;
+    vfs_msg->pid    = -1;
+
+    memset(cbuf, 0, sizeof(cbuf));
+    memset(&msgh, 0, sizeof(msgh));
+    iov.iov_base        = &msg;
+    iov.iov_len         = sizeof(msg);
+    msgh.msg_iov        = &iov;
+    msgh.msg_iovlen     = 1;
+    msgh.msg_control    = cbuf;
+    msgh.msg_controllen = sizeof(cbuf);
+
+    do {
+        nrecvd = recvmsg(sockfd, &msgh, MSG_WAITALL);
+    } while ((nrecvd < 0) & (errno == EINTR));
+    if (nrecvd != iov.iov_len) {
```

Comment:
`ucs_assert(nrecvd < iov.iov_len);`

Reply:
sure, but it should be assert() since vfs_sock does not link with UCS

Reply:
> sure, but it should be assert() since vfs_sock does not link with UCS

ok